packages/common/core-interfaces/src/jsonDeserialized.ts

```diff
@@ -0,0 +1,71 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import type { InternalUtilityTypes } from "./exposedUtilityTypes.js";
+
+/**
+ * Options for {@link JsonDeserialized}.
+ *
+ * @beta
+ */
+export interface JsonDeserializedOptions {
+	/**
+	 * A type that is managed by custom deserialization logic (beyond JSON.parse).
+	 *
+	 * The default value is `never`.
+	 */
+	Replaced: unknown;
+}
+
+/**
+ * Produces a type that results from a type `T` serialized and deserialized
+ * through JSON.
+ *
+ * @typeParam T - The type that was serialized.
+ *
+ * @remarks
+ * When used as a filter to inferred generic `T`, a compile-time error can be
+ * produced trying to assign `JsonDeserialized<T>` to `T`.
+ *
+ * Deserialized JSON never contains `undefined`, `symbol`, or function values.
+ * Object properties with values of those types are absent. So properties
+ * become optional (when there are other supported types in union) or are
+ * removed (when nothing else in union is supported).
+ * In an array, such values are replaced with `null`.
+ *
+ * `bigint` valued properties are simply removed as serialization attempts
+ * will throw.
+ *
+ * Setter and getter properties become value properties after filtering
+ * although no data will be persisted assuming those properties are backed
+ * by functions. If an implementation of getter/setter interface uses a
+ * simple data member (of supported type), that will persist.
+ *
+ * Recursive types without any required modification are preserved intact.
+ * Recursive types that require modification are unrolled a limited number of
+ * times and then further instances of recursion are replaced with
+ * {@link JsonTypeWith|JsonTypeWith<Options.Replaced>}.
+ *
+ * Under basic serialization, class instances become simple data objects that
+ * lose hidden properties and prototypes that are required for `instanceof`
+ * runtime checks.
+ *
+ * The optional 'Options.Replaced' parameter may be used to permit additional
+ * leaf types handled by custom serialization/deserialization logic.
+ *
+ * @example Typical usage
+ *
+ * ```typescript
+ * function foo<T>(): JsonDeserialized<T> { ... }
+ * ```
+ *
+ * @beta
+ */
+export type JsonDeserialized<
```

Comment:
I see reasonable (but in my opinion not particularly well motivated given the complexity and limitations) use-cases for JsonSserialized, but JsonDeserialized doesn't have a usecase I can find other than casting unvalidated deserialized data to a type it might not actually be compatible with. While I'm ok with customers doing such casts when they think its safe, them doing so does not require this type since they can just cast to their actual type.

Having us use this inside a library seems like its encapsulating a possibly incorrect cast which there is no possible way to validate into our framework, which seems like a bad idea, especially if we expose it publicly. We might have some internal use-cases for this, but I don't think we should use it in a public API, so maybe this specific type should be marked internal instead of beta?

Reply:
The way this API is factored doesn't align well with an API to do to optional validation of data (there is no runtime data format / schema). If you had an API which could do optional validation, you would never use this APIs with it. Tree would never want to use it since we have both internal and user facing schema systems supporting optional data validation and I personally wouldn't make a different design decision elsewhere to use this.

Also is there a need for this to go straight to beta? Could start at alpha.

Regardless: this is entirely a question of API design, I think FF-API really should make the call on if this should exist or not. While I'm on the side of not, it isn't my place to make that subjective API design decision.

The serialization side of this is much less problematic and I have no conceptual opposition to it existing (other than maintenance cost, API surface size and having to stabilize it if we make it non-internal), though if you have code that can safely serialize, it generally comes with an API to serialize so I don't see a use case for it in robust code but its existence as something people will be tempted to use doesn't seem as actively harmful like I consider  JsonDeserialized to be.